### ISSUE TITLE
[BUG] nn.functional.cosine_similarity returns NaN for finite fp16/fp32 inputs (overflow in dot-then-divide)

### ISSUE THREAD

**Issue by lexing-2026** [lexing-2026/TensorPlay#47]
**TensorPlay version:** 1.1.0a0+git2edaefb

**Install source:** Built from source

**Platform:** Linux x86_64

**Python version:** 3.13

**Did this work in a previous version?** Not sure

**CUDA Toolkit / driver version:** n/a (CPU)

## Minimal reproduction

```python
import numpy as np
import tensorplay as tp
import torch

# fp16: everyday magnitudes (dim sum exceeds the fp16 range)
a = np.full((4, 8), 100.0, dtype=np.float16)
print(tp.nn.functional.cosine_similarity(tp.tensor(a), tp.tensor(a), dim=1).tolist())
# -> [nan, nan, nan, nan]

print(torch.nn.functional.cosine_similarity(torch.tensor(a), torch.tensor(a), dim=1).tolist())
# -> [1.001..., 1.001..., 1.001..., 1.001...]

# fp32: same class of failure at very large but finite magnitudes
b = np.full((4, 8), 1e19, dtype=np.float32)
print(tp.nn.functional.cosine_similarity(tp.tensor(b), tp.tensor(b), dim=1).tolist())
# -> [nan, nan, nan, nan]
```

## Expected behavior

Cosine similarity of a vector with itself should be ~1.0 (exactly 1.0 for the
collinear case), i.e. no NaN for finite inputs. Verified references:

- torch CPU: `1.001` (fp16) and `0.0` (fp32 at 1e19 — itself an
  overflow artifact, see pytorch/pytorch#78064, but never NaN);
- computed in float64, both frameworks return exactly `1.0`;
- magnitude sweep: fp16 breaks for any `|x| >= ~100` with dim >= 8
  (sum of squares > 65504); fp32 breaks at `|x| >= ~1.9e19`
  (square/sum > 3.4e38).

## Actual behavior / error output

`nn.functional.cosine_similarity` **silently returns NaN** for all correct,
finite inputs above the overflow threshold. NaN then propagates through
normalization layers, attention and losses without any error or warning.

Threshold measurements (tp vs torch, dim=8):

| dtype | magnitude | tp | torch |
|---|---|---|---|
| fp16 | 100 | nan | 1.0010 |
| fp16 | 1000 | nan | 1.0000 |
| fp32 | 1e10 | 1.0 | 1.0 |
| fp32 | 1e19 | nan | 0.0 |
| fp32/fp64 | 1e30 | nan / 1.0 | 0.0 / 1.0 |

## Root cause

`tensorplay/nn/functional.py` (lines 1942-1946) overrides the op with:

```python
denom = x1.norm([dim]).mul(x2.norm([dim]))
denom = tensorplay.clamp(denom, min=eps)
return (x1 * x2).sum(dim) / denom
```

The dot product `(x1 * x2).sum(dim)` is computed in the **input dtype**: in
fp16 it overflows as soon as the per-row sum of squares exceeds 65504
(magnitude ~100 with dim >= 8), and in fp32 beyond ~1.9e19. The numerator
becomes `inf`, the denominator `inf`, and `inf / inf` yields NaN.

The composite kernel that serves this op for the C++ path
(`p10/src/backend/composite/Distance.cpp`, `cosine_similarity_native`)
already uses the safe evaluation order — normalize each input first, then dot,
and returns `0.0` (never NaN) for the same overflow — and fp64 accumulation
gives the exact 1.0. The Python override shadows it with the unsafe order.
torch's CPU kernel likewise normalizes first and accumulates the reduction in
a wider type (`acc_type`), which is why it stays NaN-free.

## Additional context

- The override at `tensorplay/nn/functional.py:1942-1946` is the only place
  with this dot-then-divide order; `nn.modules.distance.CosineSimilarity`
  forwards to `F.cosine_similarity`.
- Fix directions (either is sufficient; the first is minimal):
  1. Drop the Python override so the composite kernel (safe order) serves
     the op, or
  2. Rewrite as `(x1 / x1.norm(dim, keepdim=True) * (x2 / x2.norm(...))).sum(dim)`
     with the reduction in `acc_type` (fp32 for fp16, fp64 for fp32) per the
     upstream `acc_type` conventions, so even the 1.0 case is exact. For
     fp32, accumulate the squares in float64 (or scale each row by its
     max-abs first) to avoid the ~1.9e19 threshold entirely.
- Reproduced in-process against upstream 2.13.0+cu130; CPU only (no GPU in
  this environment, so CUDA behavior is unverified).
